distributed/scheduler.py

```diff
@@ -3013,6 +3017,24 @@ def get_task_stream(self, comm=None, start=None, stop=None, count=None):
         ts = [p for p in self.plugins if isinstance(p, TaskStreamPlugin)][0]
         return ts.collect(start=start, stop=stop, count=count)
 
+    @gen.coroutine
+    def register_init_func(self, comm, function=None):
+        """ Registers a preload function, and call it on every worker """
+        if function is None:
+            raise gen.Return({})
+
+        self.init_functions.append(loads(function))
+
+        responses = yield self.broadcast(msg=dict(op='run',
+                                                  function=function))
+        results = {}
+        for key, resp in responses.items():
+            if resp['status'] == 'OK':
+                results[key] = resp['result']
+            elif resp['status'] == 'error':
+                six.reraise(*clean_exception(**resp))
```

Comment:
It would be good to test the exception behavior

```python
with pytest.raises(ZeroDivisionError):
    yield client.register_initialization(lambda: 1 / 0)
```

Reply:
Added the test, but it don't run well. Stack trace is:
````
distributed.worker - INFO - Run out-of-band function '<lambda>'
distributed.worker - WARNING -  Run Failed
Function: <lambda>
args:     ()
kwargs:   {}
Traceback (most recent call last):
  File "/home/eh/eynardbg/git/distributed/distributed/worker.py", line 1036, in run
    result = function(*args, **kwargs)
  File "/home/eh/eynardbg/git/distributed/distributed/tests/test_worker.py", line 1273, in <lambda>
    yield c.register_init_func(lambda: 1 / 0)
ZeroDivisionError: division by zero
distributed.worker - INFO - Run out-of-band function '<lambda>'
distributed.worker - WARNING -  Run Failed
Function: <lambda>
args:     ()
kwargs:   {}
Traceback (most recent call last):
  File "/home/eh/eynardbg/git/distributed/distributed/worker.py", line 1036, in run
    result = function(*args, **kwargs)
  File "/home/eh/eynardbg/git/distributed/distributed/tests/test_worker.py", line 1273, in <lambda>
    yield c.register_init_func(lambda: 1 / 0)
ZeroDivisionError: division by zero
distributed.core - ERROR - 'Serialized' object has no attribute '__traceback__'
Traceback (most recent call last):
  File "/home/eh/eynardbg/git/distributed/distributed/core.py", line 340, in handle_comm
    result = yield result
  File "/work/ADM/hpc/eynardbg/Outils/miniconda3/envs/dask_dev/lib/python3.5/site-packages/tornado/gen.py", line 1133, in run
    value = future.result()
  File "/work/ADM/hpc/eynardbg/Outils/miniconda3/envs/dask_dev/lib/python3.5/asyncio/futures.py", line 294, in result
    raise self._exception
  File "/work/ADM/hpc/eynardbg/Outils/miniconda3/envs/dask_dev/lib/python3.5/site-packages/tornado/gen.py", line 1147, in run
    yielded = self.gen.send(value)
  File "/home/eh/eynardbg/git/distributed/distributed/scheduler.py", line 3036, in register_init_func
    six.reraise(*clean_exception(**resp))
  File "/work/ADM/hpc/eynardbg/Outils/miniconda3/envs/dask_dev/lib/python3.5/site-packages/six.py", line 691, in reraise
    if value.__traceback__ is not tb:
AttributeError: 'Serialized' object has no attribute '__traceback__'

````

Reply:
So here I will need some help, I really don't know what's wrong!

Reply:
I'll take a look

Reply:
So it's similar to the other situation with pickle.  The scheduler doesn't deserialize data by default.  Instead, it's better to pass along the results to the client and raise the exception there.  This is good for a few reasons:

1.  You want the exception to raise on the client anyway, because that is where the user is.
2.  The scheduler may not have the same versions as the client and workers (which must be the same) so deserializing the exception here may result in another error
3.  In the future the scheduler may be written in another language, so the exceptions may not make sense
4.  We tend to treat the scheduler with a greater sense of security, and so don't deserialize data here.

The object you're running into, `Serialized`, is something that holds onto serialized bytes and passes them along.  By default all of the communication channels running into the scheduler don't deserialize their data, and instead pass along these `Serialized` objects.

So you probably want to return whatever `Scheduler.broadcast` gives you and hand it back to the Client.  On the client side you'll want to handle the exception, much like is done within the `Client._run` method.  Now that we're doing more logic we may now actually need separate coroutine and blocking methods (like `_register_init_func` and `register_init_func`).  My apologies for pushing you away from those earlier.

Reply:
Thanks for the explanation, very helpful, fixed now.

Not sure I understand all the subtleties of subroutine and passing args or kwargs around, bu it will come with more practice!